### PR TITLE
🏗 Optimize root-level and task-level package installation while running `amp` tasks

### LINKED ISSUE #33141: gulp runner should handle tasks' package.json instead of the task handler itself
i.e., tasks that have their own package.json (such as `gulp visual-diff`, `gulp e2e`) should not have to perform dependency management, instead have our gulp runner itself figure out whether that's required for each task

### PR DESCRIPTION
**Pull request by rsimha**
**PR Highlights:**

- Add a step to the `amp` task runner that auto-installs task-specific subpackages if necessary
- Consolidate `updatePackages` and `updateSubpackages` in `common/update-packages.js`
- Remove `amp update-packages` now that packages are automatically updated before every other task
- Change dir to repo-root and update root packages only when a task is actually being run (skip for `--help`)
- Delete `initializeRunner()` now that there's nothing left to do in it

**Screenshots:**

**First run during local dev:**

image

**Subsequent run during local dev:**

image

**Clean run during CI:**

image

Fixes #33141

### CODE REVIEW

**Comment by alanorozco** [comment]
> auto-installs task-specific subpackages if necessary

Nice!

File: build-system/tasks/amp-task-runner.js [hunk 1]
```diff
@@ -165,7 +180,7 @@ function finalizeRunner() {
     log(red('ERROR:'), 'Unknown task', cyan(args.join(' ')));
     log('⤷ Run', cyan('amp --help'), 'for a full list of tasks.');
     log('⤷ Run', cyan('amp <task> --help'), 'for help with a specific task.');
-    process.exitCode = 1;
+    process.exit(1);
```
Comment (danielrozenberg) [inline build-system/tasks/amp-task-runner.js:183]: Why this change?

File: build-system/tasks/release/index.js [hunk 2]
```diff
@@ -121,9 +121,6 @@ async function prepareEnvironment_(outputDir, tempDir) {
   await fs.emptyDir(outputDir);
   await fs.emptyDir(tempDir);
   logSeparator_();
-
```
Comment (danielrozenberg) [inline build-system/tasks/release/index.js:124]: Remove "ensures dependencies are installed" from the jsdoc

File: build-system/tasks/visual-diff/index.js [hunk 3]
```diff
@@ -805,10 +805,6 @@ async function ensureOrBuildAmpRuntimeInTestMode_() {
  * @return {!Promise<!puppeteer.BrowserFetcher>}
  */
 async function installDependencies_() {
```
Comment (danielrozenberg) [inline build-system/tasks/visual-diff/index.js:807]: This function can now be replaced by just directly importing these packages at the top, right?

**Review by danielrozenberg — APPROVED** [review]
Approval for shared code and release + visual-diff tasks - I'll leave the other tasks to those reviewers more familiar with them

File: build-system/tasks/amp-task-runner.js (hunk 1 above)
Comment (rsimha) [reply build-system/tasks/amp-task-runner.js:183]: We started setting `exitCode` instead of directly exiting because `gulp` used to complain about unresolved promises and an uncalled `done()`. That's no longer a problem now.

File: build-system/tasks/release/index.js (hunk 2 above)
Comment (rsimha) [reply build-system/tasks/release/index.js:124]: Good catch. Done.

File: build-system/tasks/visual-diff/index.js (hunk 3 above)
Comment (rsimha) [reply build-system/tasks/visual-diff/index.js:807]: They cannot because when we load this file in `--help` mode, the dependencies are not going to be present. I've updated the JSDoc here as well.

File: build-system/tasks/amp-task-runner.js (hunk 1 above)
Comment (estherkim) [reply build-system/tasks/amp-task-runner.js:183]: IMO it's better to set the exit code and letting the process exit gracefully. Because with `process.exit()`, there could be async callbacks or other stragglers that are terminated unexpectedly.

Comment (rsimha) [reply build-system/tasks/amp-task-runner.js]: After looking at the what else is done by `commander` in this scenario, I've reverted the change and gone back to setting `exitCode`.

However, there do remain other places where a hard exit is necessary. E.g. The task runner notices that the task implementation has encountered a fatal error, and waiting for graceful exit could result in the task stalling and never exiting. In those cases, the only way out is to call `process.exit(1)`.

**Comment by rsimha** [comment]
I've manually verified the changes in the remaining tasks. Merging now.
